### PR TITLE
Handle saved payment method selections in vertical mode

### PR DESCRIPTION
**Pull request by porter-stripe**
## Summary
- Fixes a bug when you update a card brand, then exit editing mode it was not re-selected
- Fixes a bug where the stack view of the `PaymentMethodRowButton` was intercepting taps making the row button untappable for most of it's width.
- Computes the correct accessory button to show next to the saved payment method
- Shows the correct screen when tapping the accessory button (e.g. when only 1 card open the update view controller)

https://github.com/stripe/stripe-ios/assets/88012362/728c4113-7aa9-4aa1-8b25-0e63f6aa3163

## Motivation
Vert mode

## Testing
- Manual
- New unit tests

## Changelog
N/A

### CODE REVIEW

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/RightAccessoryButtonTest.swift [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    func testSingleCard_Returns_CorrectValue() {
+        // Iterating over all combinations of isFirstCardCoBranded, isCBCEligible, allowsRemovalOfLastSavedPaymentMethod, and paymentMethodRemove
+        let booleans = [false, true]
+        for isFirstCardCoBranded in booleans {
+            for isCBCEligible in booleans {
+                for allowsRemovalOfLastSavedPaymentMethod in booleans {
+                    for paymentMethodRemove in booleans {
+
+                        // We should only show the edit button if it is update-able or removable
+                        let canEdit = (isFirstCardCoBranded && isCBCEligible) || (allowsRemovalOfLastSavedPaymentMethod && paymentMethodRemove)
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/RightAccessoryButtonTest.swift:47]: Don't love this logic living in the test, but it's a tradeoff to get a very exhaustive set of tests on something that is edge-casey.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentSheetVerticalViewController.swift [hunk 2]
```diff
@@ -12,7 +12,22 @@
 import UIKit
 
 class PaymentSheetVerticalViewController: UIViewController, FlowControllerViewControllerProtocol, PaymentSheetViewControllerProtocol {
-    var selectedPaymentOption: PaymentSheet.PaymentOption?
+    // TODO make sure `selectedPaymentOption` is correct
+    var selectedPaymentOption: PaymentSheet.PaymentOption? {
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentSheetVerticalViewController.swift:16]: I want to flesh out this relationship between `PaymentSheet.PaymentOption` <> `VerticalPaymentMethodListSelection` in a later PR. Probably in the PR where we handling when we tap Link in the wallet view.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentSheetVerticalViewController.swift [hunk 3]
```diff
@@ -54,10 +73,23 @@ class PaymentSheetVerticalViewController: UIViewController, FlowControllerViewCo
...
+        )
+    }
+
     // MARK: - Initializers
 
     init(configuration: PaymentSheet.Configuration, loadResult: PaymentSheetLoader.LoadResult, isFlowController: Bool) {
-        // TODO: Deal with previousPaymentOption
+        // TODO: Deal with previousPaymentOption, default to first saved PM for now
+        if let savedPaymentMethod = loadResult.savedPaymentMethods.first {
+            self.selectedPaymentOption = .saved(paymentMethod: savedPaymentMethod, confirmParams: nil)
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentSheetVerticalViewController.swift:91]: Ah shoot I've implemented `selectedPaymentOption` as well, and in a different way (it's computed)...I'll probably let this PR through and then refactor your implementation / deal with the merge conflicts.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentMethodsViewController.swift [hunk 4]
```diff
@@ -298,6 +304,7 @@ extension VerticalSavedPaymentMethodsViewController: UpdateCardViewControllerDel
         // Create the new button
         let newButton = PaymentMethodRowButton(paymentMethod: updatedPaymentMethod, appearance: configuration.appearance)
         newButton.delegate = self
+        newButton.state = oldButton.previousSelectedState // save the previous state from the old button too
         newButton.state = oldButton.state
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentMethodsViewController.swift:308]: Typo?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Vertical Main Screen/RightAccessoryButton.swift [hunk 5]
```diff
@@ -83,15 +78,43 @@ extension RowButton {
...
+    ///   - isFirstCardCoBranded: True if the first saved payment method is a co-branded card, false otherwise
+    ///   - isCBCEligible: True if the merchant is eligible for card brand choice, false otherwise
+    ///   - allowsRemovalOfLastSavedPaymentMethod: True if we can remove the last saved payment method, false otherwise
+    ///   - paymentMethodRemove: True if removing payment methods is enabled, false otherwise
+    /// - Returns: 'AccessoryType.viewMore' if more than one payment method is saved, 'AccessoryType.edit' if only one payment method exists and it can either be updated or removed, and 'nil' otherwise.
+    static func getAccessoryButtonType(savedPaymentMethodsCount: Int,
+                                       isFirstCardCoBranded: Bool,
+                                       isCBCEligible: Bool,
+                                       allowsRemovalOfLastSavedPaymentMethod: Bool,
+                                       paymentMethodRemove: Bool) -> AccessoryType? {
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Vertical Main Screen/RightAccessoryButton.swift:109]: nit: rename to `allowsPaymentMethodRemoval`

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentSheetVerticalViewController.swift [hunk 6]
```diff
@@ -114,6 +153,35 @@ class PaymentSheetVerticalViewController: UIViewController, FlowControllerViewCo
...
+        // If we have only one row in the vertical list and it collects user input, display the form instead of the payment method list.
+        let firstPaymentMethodType = paymentMethodTypes[0]
+        // TODO: Handle offerSaveToLinkWhenSupported, previousCustomerInput, delegate
+        let pmFormVC = PaymentMethodFormViewController(type: firstPaymentMethodType, intent: intent, previousCustomerInput: nil, configuration: configuration, isLinkEnabled: false, delegate: self)
+        if paymentMethodListViewController.rowCount == 1 && pmFormVC.form.collectsUserInput {
+            self.paymentMethodFormViewController = pmFormVC
+            add(childViewController: pmFormVC, containerView: paymentContainerView)
+        } else {
+            add(childViewController: paymentMethodListViewController, containerView: paymentContainerView)
+        }
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentSheetVerticalViewController.swift:183]: Hmm this can't be in here, it should only be called once to determine the initial state. I'll deal with it in a follow-up PR though.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Views/VerticalPaymentMethodListView.swift [hunk 7]
```diff
@@ -31,7 +42,7 @@ class VerticalPaymentMethodListView: UIView {
         return stackView.arrangedSubviews.compactMap { $0 as? RowButton }
     }
 
-    init(savedPaymentMethod: STPPaymentMethod?, paymentMethodTypes: [PaymentSheet.PaymentMethodType], shouldShowApplePay: Bool, shouldShowLink: Bool, appearance: PaymentSheet.Appearance) {
+    init(currentSelection: VerticalPaymentMethodListSelection?, savedPaymentMethod: STPPaymentMethod?, paymentMethodTypes: [PaymentSheet.PaymentMethodType], shouldShowApplePay: Bool, shouldShowLink: Bool, rightAccessoryType: RowButton.RightAccessoryButton.AccessoryType?, appearance: PaymentSheet.Appearance) {
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Views/VerticalPaymentMethodListView.swift:45]: nit: I'd rename `rightAccessoryType` to `savedPaymentMethodAccessoryType` or something

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Views/VerticalPaymentMethodListView.swift [hunk 8]
```diff
@@ -42,14 +53,26 @@ class VerticalPaymentMethodListView: UIView {
...
+                                                                               rightAccessoryView: accessoryButton)
+            { [weak self] in
                 self?.didTap(rowButton: $0, selection: .saved(paymentMethod: savedPaymentMethod))
             }
-            // Make the saved PM the default selected
-            savedPaymentMethodButton.isSelected = true
-            currentSelection = .saved(paymentMethod: savedPaymentMethod)
+
+            // Select saved payment method button if current selection is a saved payment method or if current selection is nil
+            if case .saved(let savedPaymentMethod) = currentSelection {
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Views/VerticalPaymentMethodListView.swift:68]: Gahh I also implemented this.  Can we rename to `initialSelection` (what I named this) to make my merge conflicts a bit easier?

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/RightAccessoryButtonTest.swift (hunk 1 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/RightAccessoryButtonTest.swift:47]: Yeah I'm not sure how I feel about these tests, hah. It feels so tightly coupled to the implementation of the function it's testing. I think I'd prefer if we tested this at a higher level so it can detect bugs outside of this simple method. I dont' feel strongly right now, can follow up.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentMethodsViewController.swift (hunk 4 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentMethodsViewController.swift:308]: Technically no, but it does read funny. The second state update will preserve the first state as it's previous state. I've made it easier to read.
